### PR TITLE
Completes CLI Interactivity for Grant, Encrypt and Retrieve

### PR DESCRIPTION
**Pull request by KPrasch**
**Type of PR:**
Feature 

**Required reviews:** 
2

**What this does:**
Completes CLI interactivity and parameter collection for CLI usage of Alice, Bob, and Enrico.

The following usages are now supported, all flags are optional:

```
nucypher alice grant
nucypher enrico encrypt
nucypher bob retrieve
```

**Why it's needed:**
Improves usability and legibility of grant and retrieve while using the CLI.  This is particularly impactful for application developers  and demo presenters.

### CODE REVIEW

File: nucypher/blockchain/eth/interfaces.py [hunk 1]
```diff
@@ -584,6 +585,10 @@ def sign_and_broadcast_transaction(self,
         #
 
         try:  # TODO: Handle block confirmation exceptions
+            if confirmations:
+                emitter.message(f'Waiting for {confirmations} confirmations', color='yellow')
+            else:
+                emitter.message(f'Waiting {self.TIMEOUT} seconds for receipt', color='yellow')
```
Comment (derekpierre) [inline nucypher/blockchain/eth/interfaces.py:591]: I don't think confirmations and timeout are mutually exclusive. It's a "keep trying until one condition, sufficient confirmations or timeout, happens first", no? If so. we may need to reword.

Perhaps something like:
```suggestion
            waiting_for = 'receipt'
            if confirmations:
                waiting_for = f'{confirmations} confirmations'
            emitter.message(f'Waiting {self.TIMEOUT} seconds for {waiting_for}', color='yellow')
```

File: nucypher/characters/control/specifications/bob.py [hunk 2]
```diff
@@ -42,23 +42,30 @@ class JoinPolicy(BaseSchema):  #TODO:  this doesn't have a cli implementation
 
 class Retrieve(BaseSchema):
     label = fields.Label(
-        required=True, load_only=True,
-        click=options.option_label(required=True))
+        required=True,
+        load_only=True,
+        click=options.option_label(required=False)
```
Comment (derekpierre) [inline nucypher/characters/control/specifications/bob.py:47]: Presumably, there are no side effects of this change, and the only effect is what the PR intended i.e. that Bob will be prompted for the label if not explicitly specified.

File: nucypher/cli/commands/alice.py [hunk 3]
```diff
@@ -483,6 +484,13 @@ def grant(general_config,
         bob_encrypting_key = card.encrypting_key.hex()
         bob_verifying_key = card.verifying_key.hex()
 
+    if not bob_encrypting_key or bob_verifying_key:
+        emitter.message("*Caution: Only enter public keys*")
+    if not bob_encrypting_key:
+        bob_encrypting_key = click.prompt("Enter Bob's encrypting key")
+    if not bob_verifying_key:
+        bob_verifying_key = click.prompt("Enter Bob verifying key")
+
```
Comment (derekpierre) [inline nucypher/cli/commands/alice.py:493]: Do we need to be careful here of `force` is true, or is that case handled earlier?

File: nucypher/cli/commands/alice.py [hunk 4]
```diff
@@ -500,14 +508,12 @@ def grant(general_config,
...
     if not force and not expiration:
-        if ALICE.duration_periods:
-            # TODO: use a default in days or periods?
-            expiration = maya.now() + timedelta(days=ALICE.duration_periods)  # default
-            if not click.confirm(f'Use default policy duration (expires {expiration})?'):
-                expiration = click.prompt('Enter policy expiration datetime', type=click.DateTime())
-        else:
-            # No policy duration default default available; Go interactive
-            expiration = click.prompt('Enter policy expiration datetime', type=click.DateTime())
+        expiration_prompt = 'Enter policy expiration (Y-M-D H:M:S)'
```
Comment (derekpierre) [inline nucypher/cli/commands/alice.py:511]: I know we allowed `click.DateTime()` before which allowed for formats `["%Y-%m-%d", "%Y-%m-%dT%H:%M:%S", "%Y-%m-%d %H:%M:%S"]`, but can we actually do anything more granular than day i.e. hours, minutes, seconds? If we can't, then we should explicitly specify the format `"%Y-%m-%d"` to click.DateTime, and only prompt for `Y-m-d` to be specified.

File: nucypher/cli/commands/enrico.py [hunk 5]
```diff
@@ -63,25 +63,38 @@ def run(general_config, policy_encrypting_key, dry_run, http_port):
...
     ENRICO = _create_enrico(emitter, policy_encrypting_key)
-    if not (bool(message) ^ bool(file)):
-        emitter.error(f'Pass either --message or --file. Got {message}, {file}.')
+    if message and file:
+        emitter.error(f'Pass either --message or --file, not both.')
         raise click.Abort
 
+    if not message:
+        message = click.prompt('Enter plaintext to encrypt', type=click.STRING)
+
```
Comment (derekpierre) [inline nucypher/cli/commands/enrico.py:88]: confused here - what if `file` was passed; doesn't this nullify the original XOR check?

File: nucypher/cli/utils.py [hunk 6]
```diff
@@ -101,6 +101,7 @@ def make_cli_character(character_config,
     if character_config.federated_only:
         emitter.message(FEDERATED_WARNING, color='yellow')
 
+    emitter.message(f"Loaded {CHARACTER.__class__.__name__} {CHARACTER.checksum_address} ({CHARACTER.domain})", color='green')
```
Comment (derekpierre) [inline nucypher/cli/utils.py:104]: 👍

File: nucypher/cli/commands/alice.py [hunk 7]
```diff
@@ -525,14 +531,15 @@ def grant(general_config,
     # Policy Value
     policy_value_provided = bool(value) or bool(rate)
     if not ALICE.federated_only and not policy_value_provided:
-        rate = ALICE.default_rate  # TODO #1709 - Fine tuning and selection of default rates
+        # TODO #1709 - Fine tuning and selection of default rates
+        rate = ALICE.default_rate or Web3.toWei(50, 'gwei')  # TODO: Unhardcode, workaround for lynx/ibex testnet.
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/commands/alice.py:535]: Current this value is 0 in lynx, and I wanted to show some number here for a demo.  What is better to show in this case?

File: nucypher/characters/control/specifications/bob.py (hunk 2 above)
Comment (KPrasch) [reply nucypher/characters/control/specifications/bob.py:47]: Yes, that's correct, the Click options need be be optional to allow the fallback to interactive collection.

File: nucypher/cli/commands/alice.py (hunk 4 above)
Comment (vepkenez) [reply nucypher/cli/commands/alice.py:511]: Yes.  If find these precise format requirements to be very frictional.

Comment (KPrasch) [reply nucypher/cli/commands/alice.py:511]: > can we actually do anything more granular than day i.e. hours, minutes, seconds?

Yes, epoch seconds is used at the contract-level.

File: nucypher/cli/commands/enrico.py (hunk 5 above)
Comment (KPrasch) [reply nucypher/cli/commands/enrico.py:88]: Unsure - this is more or less an "application level" decision.  What's a sensible default behavior here?

Comment (derekpierre) [reply nucypher/cli/commands/enrico.py:88]: I guess my thought was that the original code ensured that only one of them is set by using the XOR - `if not (bool(message) ^ bool(file)):`. In this change, we ensure that both aren't set, `if message and file` but we don't know which one, `file` or `message` is actually set, or if they are set at all.

Subsequently, you check on L86 whether `message` is not set, which could mean that either `file` is or is not set`. I think to do what you are trying to do here you might have to do something like:

```suggestion
    if not file and not message:
        message = click.prompt('Enter plaintext to encrypt', type=click.STRING)

```

File: nucypher/cli/commands/alice.py (hunk 4 above)
Comment (derekpierre) [reply nucypher/cli/commands/alice.py:511]: Interesting. I guess in that case there is payment for periods and any part thereof i.e. payment for fees is period granular (effectively UTC day), but enaction of a policy is second granularity (@vzotova do I have this right?).

File: nucypher/cli/commands/enrico.py (hunk 5 above)
Comment (KPrasch) [reply nucypher/cli/commands/enrico.py:88]: > I guess my thought was that the original code ensured that only one of them is set by using the XOR

It doesn't play nicely with the interactivity -  When might this check be effective if we're prompting for manual input?

>  I think to do what you are trying to do here you might have to do something like

I'm liking this simple approach here, with the default behavior focused around encrypting messages (good for first-time users, hobbyist, demoers).  `--file` can be added if needed.

```suggestion
    if not message or file:
        message = click.prompt('Enter plaintext to encrypt', type=click.STRING)

```

File: nucypher/cli/commands/alice.py (hunk 4 above)
Comment (KPrasch) [reply nucypher/cli/commands/alice.py:511]: Policy rates are calculated on a per-period basis, but retain their original timeatamps:

https://github.com/nucypher/nucypher/blob/main/nucypher/blockchain/eth/sol/source/contracts/PolicyManager.sol#L239

File: nucypher/cli/commands/alice.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (vzotova) [reply nucypher/cli/commands/alice.py:535]: I'm for setting default value on `lynx`

File: nucypher/cli/commands/enrico.py (hunk 5 above)
Comment (KPrasch) [reply nucypher/cli/commands/enrico.py]: :heavy_check_mark:

File: nucypher/cli/commands/alice.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/commands/alice.py]: :heavy_check_mark: - Handled 0 value for rate in 55f2ede773390c6197ca4f830029a719a78f6cce.  This is an exceptional case, so maybe it;s better to leave out this commit.

File: nucypher/cli/commands/alice.py (hunk 3 above)
Comment (KPrasch) [reply nucypher/cli/commands/alice.py]: :heavy_check_mark:

File: nucypher/cli/actions/collect.py [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+        if not force:
+            default_gwei = Web3.fromWei(rate, 'gwei')  # wei -> gwei
+            prompt = "Confirm rate of {node_rate} gwei * {n} nodes ({period_rate} gwei per period)?"
+
+            if not click.confirm(prompt.format(node_rate=default_gwei, period_rate=default_gwei * n, n=n), default=True):
+                interactive_rate = click.prompt('Enter rate per period in gwei', type=GWEI)
+                # TODO: Interactive rate sampling & validation (#1709)
+                interactive_prompt = prompt.format(node_rate=interactive_rate, period_rate=interactive_rate * n, n=n)
+                click.confirm(interactive_prompt, default=True, abort=True)
+                rate = Web3.toWei(interactive_rate, 'gwei')  # gwei -> wei
```
Comment (cygnusv) [inline nucypher/cli/actions/collect.py:146]: Does this solve the problem I found?
```
Confirm rate of 50 gwei * 3 (150 gwei per period)? [Y/n]: n
Enter rate per period in gwei: 60
Confirm rate of 50000000000 gwei * 3 (150000000000 gwei per period)? [Y/n]:
```

**Review by cygnusv — APPROVED** [review]
Tested live on Ethereum Denver...Works like a charm! Thanks for this @KPrasch !
The collect and confirm modules are a great approach to keep the code tidy. The CLI code looks much better now.

File: nucypher/cli/actions/collect.py [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+        force: bool,
+        card_identifier: str,
+        bob_encrypting_key: str,
+        bob_verifying_key: str
+        ) -> PublicKeys:
+    """helper function for collecting Bob's public keys interactively in the Alice CLI."""
+
+    # From Bob card
+    if card_identifier and any((bob_encrypting_key, bob_verifying_key)):
+        message = '--bob cannot be used with --bob-encrypting-key or --bob-verifying key'
```
Comment (derekpierre) [inline nucypher/cli/actions/collect.py:67]: Feels like a little leakage here...`--bob` as it relates to `card_identifier`

We could leave it or move verifications to the originating source before we call `collect_bob_public_keys()`. Something to think about...

File: nucypher/cli/actions/collect.py [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+
+    if card_identifier:
+        public_keys = collect_keys_from_card(
+            emitter=emitter,
+            card_identifier=card_identifier,
+            force=force)
+        return public_keys
+
+    # From hex public keys
+    if not bob_encrypting_key or bob_verifying_key:
```
Comment (derekpierre) [inline nucypher/cli/actions/collect.py:78]:
```suggestion
    if not bob_encrypting_key or not bob_verifying_key:
```

**Comment by vepkenez** [comment]
> Would you mind if I move all this validation logic into the Serializer since that's the whole point of that?

Or we should get rid of the serializers since this becomes more difficult if validation logic is pepperer all over the place... but I still think it's useful to validate at the "model" level than on the CLI interface level since much of this logic still applies.

File: nucypher/cli/actions/collect.py (hunk 8 above)
Comment (KPrasch) [reply nucypher/cli/actions/collect.py:128]: :+1:  Yes it does.

File: nucypher/cli/commands/alice.py [hunk 11]
```diff
@@ -525,14 +531,15 @@ def grant(general_config,
...
         if not force:
             default_gwei = Web3.fromWei(rate, 'gwei')
-            prompt = "Confirm rate of {node_rate} gwei ({total_rate} gwei per period)?"
-            if not click.confirm(prompt.format(node_rate=default_gwei, total_rate=default_gwei*n), default=True):
+            prompt = "Confirm rate of {node_rate} gwei * {n} ({total_rate} gwei per period)?"
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/commands/alice.py]: :heavy_check_mark:

File: nucypher/cli/actions/collect.py (hunk 9 above)
Comment (KPrasch) [reply nucypher/cli/actions/collect.py]: :heavy_check_mark:

File: nucypher/cli/actions/collect.py (hunk 10 above)
Comment (KPrasch) [reply nucypher/cli/actions/collect.py]: :heavy_check_mark:

**Comment by KPrasch** [comment]
> I still think it's useful to validate at the "model" level than on the CLI interface level since much of this logic still applies to other interaction contexts.

@vepkenez Yes, exactly. In fact I think we have developed just the right amount of granularity in validation.  Serializers in `character.control.specifications` are performing an essential role by validating the grant quest before it is passed to the character, on the "model layer" as you mentioned.  The CLI validators relocated in 005e819 (at the "application layer") check flag combinations that include fields that are not part of `Alice.grant` data model, like `--card.`   There may be an abstraction leak by deriving the click fields in the serializes, but it's doing a fine job - no reason to rework it IMO.

File: nucypher/cli/actions/collect.py [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+        public_keys = collect_keys_from_card(
+            emitter=emitter,
+            card_identifier=card_identifier,
+            force=force)
+        return public_keys
+
+    if not bob_encrypting_key:
+        bob_encrypting_key = click.prompt("Enter Bob's encrypting key")
+    if not bob_verifying_key:
+        bob_verifying_key = click.prompt("Enter Bob verifying key")
```
Comment (derekpierre) [inline nucypher/cli/actions/collect.py:75]:
```suggestion
        bob_verifying_key = click.prompt("Enter Bob's verifying key")
```

File: nucypher/cli/commands/enrico.py [hunk 13]
```diff
@@ -63,25 +63,38 @@ def run(general_config, policy_encrypting_key, dry_run, http_port):
...
     """Encrypt a message under a given policy public key."""
 
-    # Setup
     emitter = setup_emitter(general_config=general_config, banner=policy_encrypting_key)
+
+    if ipfs:
+        try:
+            import ipfshttpclient
+        except ImportError:
+            raise ImportError("IPFS HTTP client not installed.  Run pip install ipfshttpclient then try again.")
```
Comment (derekpierre) [inline nucypher/cli/commands/enrico.py:72]:
```suggestion
            raise ImportError("IPFS HTTP client not installed.  Run 'pip install ipfshttpclient' then try again.")
```

File: nucypher/cli/commands/enrico.py [hunk 14]
```diff
@@ -63,25 +63,38 @@ def run(general_config, policy_encrypting_key, dry_run, http_port):
...
+
+        # Connect to IPFS before proceeding
+        ipfs_client = ipfshttpclient.connect(ipfs)
+        emitter.message(f"Connected to IPFS Gateway {ipfs}")
+
+    if not policy_encrypting_key:
+        policy_encrypting_key = click.prompt("Enter policy public key", type=click.STRING)
```
Comment (derekpierre) [inline nucypher/cli/commands/enrico.py:79]:
```suggestion
        policy_encrypting_key = click.prompt("Enter policy encrypting key", type=click.STRING)
```

**Review by derekpierre — APPROVED** [review]
🎸 - looks good! A few minor suggestions.

**Review by vepkenez — APPROVED** [review]

**Pull request merged**